<h1>02 数据探索<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#工具包导入&amp;数据读取" data-toc-modified-id="工具包导入&amp;数据读取-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>工具包导入&amp;数据读取</a></span><ul class="toc-item"><li><span><a href="#工具包导入" data-toc-modified-id="工具包导入-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>工具包导入</a></span></li><li><span><a href="#读取数据" data-toc-modified-id="读取数据-1.2"><span class="toc-item-num">1.2&nbsp;&nbsp;</span>读取数据</a></span></li></ul></li><li><span><a href="#数据初步分析" data-toc-modified-id="数据初步分析-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>数据初步分析</a></span><ul class="toc-item"><li><span><a href="#数据边界" data-toc-modified-id="数据边界-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>数据边界</a></span></li><li><span><a href="#训练集与测试集的相关性" data-toc-modified-id="训练集与测试集的相关性-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>训练集与测试集的相关性</a></span></li><li><span><a href="#数据的分布" data-toc-modified-id="数据的分布-2.3"><span class="toc-item-num">2.3&nbsp;&nbsp;</span>数据的分布</a></span></li></ul></li><li><span><a href="#数据处理及分析" data-toc-modified-id="数据处理及分析-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>数据处理及分析</a></span><ul class="toc-item"><li><span><a href="#将特征数值化" data-toc-modified-id="将特征数值化-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>将特征数值化</a></span></li><li><span><a href="#数据分析" data-toc-modified-id="数据分析-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>数据分析</a></span></li><li><span><a href="#箱线图查看分布" data-toc-modified-id="箱线图查看分布-3.3"><span class="toc-item-num">3.3&nbsp;&nbsp;</span>箱线图查看分布</a></span></li><li><span><a href="#直方图和QQ图" data-toc-modified-id="直方图和QQ图-3.4"><span class="toc-item-num">3.4&nbsp;&nbsp;</span>直方图和QQ图</a></span></li><li><span><a href="#对比分布" data-toc-modified-id="对比分布-3.5"><span class="toc-item-num">3.5&nbsp;&nbsp;</span>对比分布</a></span></li><li><span><a href="#可视化线性关系" data-toc-modified-id="可视化线性关系-3.6"><span class="toc-item-num">3.6&nbsp;&nbsp;</span>可视化线性关系</a></span></li></ul></li></ul></div>

# 工具包导入&数据读取

##  工具包导入

In [1]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
from datetime import date
import datetime as dt
from scipy import stats
import warnings
warnings.filterwarnings("ignore")
 
%matplotlib inline

## 读取数据

In [2]:
off_train = pd.read_csv('data/ccf_线下_stage1_train.csv',keep_default_na=True)
off_train.columns=['user_id','merchant_id','coupon_id','discount_rate','distance','date_received','date']
off_test = pd.read_csv('data/ccf_线下_stage1_test_revised.csv',keep_default_na=True)
off_test.columns = ['user_id','merchant_id','coupon_id','discount_rate','distance','date_received']
on_train = pd.read_csv('data/ccf_线上_stage1_train.csv',keep_default_na=True)
on_train.columns = ['user_id','merchant_id','action','coupon_id','discount_rate','date_received','date']
off_train[['user_id','merchant_id','coupon_id']]=off_train[['user_id','merchant_id','coupon_id']].astype(str)
off_test[['user_id','merchant_id','coupon_id']]=off_test[['user_id','merchant_id','coupon_id']].astype(str)
on_train[['user_id','merchant_id','coupon_id']]=on_train[['user_id','merchant_id','coupon_id']].astype(str)

## 初步探查

In [3]:
#线下训练数据
off_train.head()

user_id merchant_id coupon_id discount_rate  distance  date_received  \
0  1439408        2632       nan           NaN       0.0            NaN   
1  1439408        4663   11002.0        150:20       1.0     20160528.0   
2  1439408        2632    8591.0          20:1       0.0     20160217.0   
3  1439408        2632    1078.0          20:1       0.0     20160319.0   
4  1439408        2632    8591.0          20:1       0.0     20160613.0   

         date  
0  20160217.0  
1         NaN  
2         NaN  
3         NaN  
4         NaN

In [4]:
#线上训练数据
on_train.head()

user_id merchant_id  action  coupon_id discount_rate  date_received  \
0  13740231       18907       2  100017492        500:50     20160513.0   
1  13740231       34805       1        nan           NaN            NaN   
2  14336199       18907       0        nan           NaN            NaN   
3  14336199       18907       0        nan           NaN            NaN   
4  14336199       18907       0        nan           NaN            NaN   

         date  
0         NaN  
1  20160321.0  
2  20160618.0  
3  20160618.0  
4  20160618.0

In [5]:
#测试数据
off_test.head()

user_id merchant_id coupon_id discount_rate  distance  date_received
0  4129537         450      9983          30:5       1.0       20160712
1  6949378        1300      3429          30:5       NaN       20160706
2  2166529        7113      6928        200:20       5.0       20160727
3  2166529        7113      1808        100:10       5.0       20160727
4  6172162        7605      6500          30:1       2.0       20160708

In [6]:
#线下训练数据
off_train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1754884 entries, 0 to 1754883
Data columns (total 7 columns):
 #   Column         Dtype  
---  ------         -----  
 0   user_id        object 
 1   merchant_id    object 
 2   coupon_id      object 
 3   discount_rate  object 
 4   distance       float64
 5   date_received  float64
 6   date           float64
dtypes: float64(3), object(4)
memory usage: 93.7+ MB


In [9]:
#线上训练数据
on_train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 11429826 entries, 0 to 11429825
Data columns (total 7 columns):
 #   Column         Dtype  
---  ------         -----  
 0   user_id        object 
 1   merchant_id    object 
 2   action         int64  
 3   coupon_id      object 
 4   discount_rate  object 
 5   date_received  float64
 6   date           float64
dtypes: float64(2), int64(1), object(4)
memory usage: 610.4+ MB


In [10]:
#测试集
off_test.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 113640 entries, 0 to 113639
Data columns (total 6 columns):
 #   Column         Non-Null Count   Dtype  
---  ------         --------------   -----  
 0   user_id        113640 non-null  object 
 1   merchant_id    113640 non-null  object 
 2   coupon_id      113640 non-null  object 
 3   discount_rate  113640 non-null  object 
 4   distance       101576 non-null  float64
 5   date_received  113640 non-null  int64  
dtypes: float64(1), int64(1), object(4)
memory usage: 5.2+ MB


distance 列有 11,064 个缺失值（113,640 - 101,576）。这表明大约 10% 的记录缺少距离信息。

In [11]:
#线下训练数据
off_train.describe()

distance  date_received          date
count  1.648881e+06   1.053282e+06  7.769840e+05
mean   2.361636e+00   2.016032e+07  2.016040e+07
std    3.483974e+00   1.771416e+02  1.621500e+02
min    0.000000e+00   2.016010e+07  2.016010e+07
25%    0.000000e+00   2.016013e+07  2.016031e+07
50%    0.000000e+00   2.016032e+07  2.016042e+07
75%    3.000000e+00   2.016051e+07  2.016052e+07
max    1.000000e+01   2.016062e+07  2.016063e+07

1. 数据分布
- distance（距离）：均值约为 2.36，表明平均距离不是特别远。标准差约为 3.48，说明距离的分布有一定的离散性。最小值为 0，最大值为 10，说明距离的范围从 0 到 10。25%、中位数（50%）和 75% 的值都是 0，表明大部分数据点的距离为 0，表示user经常活动的地点离该merchant的最近门店距离低于500米。
- date_received（领取优惠券日期）和 date（消费日期）：领取优惠券日期和消费日期它们的均值、标准差、最小值和最大值非常接近，表明它们的分布几乎相同。似乎是，用户拿到优惠卷就会立即消费？
2. 数据完整性
- distance 没有缺失值，完整性较好。
- date_received 列有 1,053,282 条记录，而 date 列有 776,984 条记录，表明存在缺失值。

In [12]:
#线上训练数据
on_train.describe()

action  date_received          date
count  1.142983e+07   8.723570e+05  1.077393e+07
mean   2.348193e-01   2.016037e+07  2.016039e+07
std    5.426314e-01   1.560031e+02  1.732371e+02
min    0.000000e+00   2.016010e+07  2.016010e+07
25%    0.000000e+00   2.016023e+07  2.016023e+07
50%    0.000000e+00   2.016040e+07  2.016041e+07
75%    0.000000e+00   2.016051e+07  2.016052e+07
max    2.000000e+00   2.016062e+07  2.016063e+07

In [7]:
##测试集
off_test.describe()

distance  date_received
count  101576.000000   1.136400e+05
mean        2.328040   2.016072e+07
std         3.260755   9.019508e+00
min         0.000000   2.016070e+07
25%         0.000000   2.016071e+07
50%         1.000000   2.016072e+07
75%         3.000000   2.016072e+07
max        10.000000   2.016073e+07

In [8]:
#线下训练数据
off_train.isnull().sum()

user_id               0
merchant_id           0
coupon_id             0
discount_rate    701602
distance         106003
date_received    701602
date             977900
dtype: int64

In [13]:
#线上训练数据
on_train.isnull().sum()

user_id                 0
merchant_id             0
action                  0
coupon_id               0
discount_rate    10557469
date_received    10557469
date               655898
dtype: int64

In [14]:
#测试集
off_test.isnull().sum()

user_id              0
merchant_id          0
coupon_id            0
discount_rate        0
distance         12064
date_received        0
dtype: int64

# 数据初步分析

## 数据边界

对数据日期范围情况进行初步探索。

In [21]:
#重新读取数据，将null原样保持，方便处理
off_train = pd.read_csv('data/ccf_线下_stage1_train.csv',keep_default_na=False)
off_train.columns=['user_id','merchant_id','coupon_id','discount_rate','distance','date_received','date']
off_test = pd.read_csv('data/ccf_线下_stage1_test_revised.csv',keep_default_na=False)
off_test.columns = ['user_id','merchant_id','coupon_id','discount_rate','distance','date_received']
on_train = pd.read_csv('data/ccf_线上_stage1_train.csv',keep_default_na=False)
on_train.columns = ['user_id','merchant_id','action','coupon_id','discount_rate','date_received','date']
off_train[['user_id','merchant_id','coupon_id']]=off_train[['user_id','merchant_id','coupon_id']].astype(str)
off_test[['user_id','merchant_id','coupon_id']]=off_test[['user_id','merchant_id','coupon_id']].astype(str)
on_train[['user_id','merchant_id','coupon_id']]=on_train[['user_id','merchant_id','coupon_id']].astype(str)

In [ ]:
print('线下 train date_received')
print(off_train[off_train['date_received']!='null']['date_received'].min())
print(off_train[off_train['date_received']!='null']['date_received'].max())

print('线上 train date_received')
print(on_train[on_train['date_received']!='null']['date_received'].min())
print(on_train[on_train['date_received']!='null']['date_received'].max())

print('test date_received')
print(off_test['date_received'].min())
print(off_test['date_received'].max())

offline train date_received
20160101
20160615
online train date_received
20160101
20160615
test date_received
20160701
20160731


In [ ]:
print('线下 train date')
print(off_train[off_train['date']!='null']['date'].min())
print(off_train[off_train['date']!='null']['date'].max())

print('线上 train date')
print(on_train[on_train['date']!='null']['date'].min())
print(on_train[on_train['date']!='null']['date'].max())

offline train date
20160101
20160630
online train date
20160101
20160630


通过探索可以发现训练数据的用券数据是到6月30日，而领券日期并不是到6月30日，而是到6月15日，这在设计滑窗结构的时候需要注意。

## 训练集与测试集的相关性

在机器学习比赛及实际应用的时候，不同数据的价值是不一样的，同时也并不是所有得到的数据都是有用的。对测试集和训练集数据的重合情况进行探查对于后续的特征构建的思路有很大的指导作用。

### 训练集中user_id与测试集的重合度

In [30]:
#查看线上 线下 训练集的 user_id与测试集的重合度
off_train_user=off_train[['user_id']].copy().drop_duplicates()
off_test_user=off_test[['user_id']].copy().drop_duplicates()
on_train_user=on_train[['user_id']].copy().drop_duplicates()
print ('线下 训练集用户ID数量')
print (off_train_user.user_id.count())
print ('线上 训练集用户ID数量')
print (on_train_user.user_id.count())
print ('线下 测试集用户ID数量')
print (off_test_user.user_id.count())

线下 训练集用户ID数量
539438
线上 训练集用户ID数量
762858
线下 测试集用户ID数量
76309


In [ ]:
off_train_user['off_train_flag']=1
off_merge=off_test_user.merge(off_train_user,on='user_id', how="left").reset_index().fillna(0)
off_merge

index  user_id  off_train_flag
0          0  4129537             1.0
1          1  6949378             1.0
2          2  2166529             1.0
3          3  6172162             1.0
4          4  4005121             1.0
...      ...      ...             ...
76304  76304  4451837             1.0
76305  76305  5828093             1.0
76306  76306  6626813             1.0
76307  76307  4547069             1.0
76308  76308  6675965             1.0

[76309 rows x 3 columns]

In [31]:
print ('线下 训练集用户与测试集用户重复数量')
print(off_merge['off_train_flag'].sum())
print ('线下 训练集用户与测试集重复用户在总测试集用户中的占比')
print(off_merge['off_train_flag'].sum()/off_merge['off_train_flag'].count())

线下 训练集用户与测试集用户重复数量
76307.0
线下 训练集用户与测试集重复用户在总测试集用户中的占比
0.9999737907717308


In [32]:
on_train_user['on_train_flag']=1
on_merge=off_test_user.merge(on_train_user,on='user_id', how="left").reset_index().fillna(0)
on_merge
print ('线上 训练集用户与测试集用户重复数量')
print(on_merge['on_train_flag'].sum())
print ('线上 训练集用户与测试集重复用户在总测试集用户中的占比')
print(on_merge['on_train_flag'].sum()/on_merge['on_train_flag'].count())

线上 训练集用户与测试集用户重复数量
43155.0
线上 训练集用户与测试集重复用户在总测试集用户中的占比
0.5655296229802513


###  训练集中merchant_id与测试集的重合度

In [33]:
#查看线上 线下 训练集的 merchant_id 与测试集的重合度
off_train_merchant=off_train[['merchant_id']].copy().drop_duplicates()
off_test_merchant=off_test[['merchant_id']].copy().drop_duplicates()
on_train_merchant=on_train[['merchant_id']].copy().drop_duplicates() 
print ('线下 训练集商家ID数量')
print (off_train_merchant.merchant_id.count())
print ('线上 训练集商家ID数量')
print (on_train_merchant.merchant_id.count())
print ('线下 测试集商家ID数量')
print (off_test_merchant.merchant_id.count())

线下 训练集商家ID数量
8415
线上 训练集商家ID数量
7999
线下 测试集商家ID数量
1559


In [34]:
off_train_merchant['off_train_flag']=1
off_merge=off_test_merchant.merge(off_train_merchant,on='merchant_id', how="left").reset_index().fillna(0)
print ('线下 训练集商家与测试集商家重复数量')
print(off_merge['off_train_flag'].sum())
print ('线下 训练集商家与测试集重复商家在总测试集商家中的占比')
print(off_merge['off_train_flag'].sum()/off_merge['off_train_flag'].count())

线下 训练集商家与测试集商家重复数量
1558.0
线下 训练集商家与测试集重复商家在总测试集商家中的占比
0.9993585631815266


In [35]:
on_train_merchant['on_train_flag']=1
on_merge=off_test_merchant.merge(on_train_merchant,on='merchant_id', how="left").reset_index().fillna(0)
print ('线上 训练集商家与测试集商家重复数量')
print(on_merge['on_train_flag'].sum())
print ('线上 训练集商家与测试集重复商家在总测试集商家中的占比')
print(on_merge['on_train_flag'].sum()/on_merge['on_train_flag'].count())

线上 训练集商家与测试集商家重复数量
0.0
线上 训练集商家与测试集重复商家在总测试集商家中的占比
0.0


### 训练集中coupon_id与测试集的重合度

In [36]:
#查看线上 线下 训练集的 coupon_id 与测试集的重合度
off_train_coupon=off_train[['coupon_id']].copy().drop_duplicates()
off_test_coupon=off_test[['coupon_id']].copy().drop_duplicates()
on_train_coupon=on_train[['coupon_id']].copy().drop_duplicates()
print ('线下 训练集优惠券ID数量')
print (off_train_coupon.coupon_id.count())
print ('线上 训练集优惠券ID数量')
print (on_train_coupon.coupon_id.count())
print ('线下 测试集优惠券ID数量')
print (off_test_coupon.coupon_id.count())

线下 训练集优惠券ID数量
9739
线上 训练集优惠券ID数量
27748
线下 测试集优惠券ID数量
2050


In [37]:
off_train_coupon['off_train_flag']=1
off_merge=off_test_coupon.merge(off_train_coupon,on='coupon_id', how="left").reset_index().fillna(0)
print ('线下 训练集优惠券与测试集优惠券重复数量')
print(off_merge['off_train_flag'].sum())
print ('线下 训练集优惠券与测试集重复优惠券在总测试集优惠券中的占比')
print(off_merge['off_train_flag'].sum()/off_merge['off_train_flag'].count())

线下 训练集优惠券与测试集优惠券重复数量
1118.0
线下 训练集优惠券与测试集重复优惠券在总测试集优惠券中的占比
0.5453658536585366


In [38]:
on_train_coupon['on_train_flag']=1
on_merge=off_test_coupon.merge(on_train_coupon,on='coupon_id', how="left").reset_index().fillna(0)
print ('线上 训练集优惠券与测试集优惠券重复数量')
print(on_merge['on_train_flag'].sum())
print ('线上 训练集优惠券与测试集重复优惠券在总测试集优惠券中的占比')
print(on_merge['on_train_flag'].sum()/on_merge['on_train_flag'].count())

线上 训练集优惠券与测试集优惠券重复数量
0.0
线上 训练集优惠券与测试集重复优惠券在总测试集优惠券中的占比
0.0


对用户，商家，优惠券在训练集和测试集的重合情况进行探索发现：
- 测试集的 用户ID 与 线下训练集 重复占比0.999以上，与线上训练集重复占比0.565。
- 测试集的 商家ID 与 线下训练集 重复占比0.999以上，与线上训练集没有重复。
- 测试集的 优惠券ID 与 训练集 都没有重复。

结论：线上数据价值比较低，后续特征提取将以 线下训练集 为主。在提取 优惠券 统计特征的时候不能通过ID进行合并。在后续可视化分析中将主要在线下训练集及测试集之间进行。

## 数据的分布

In [ ]:
off_train['discount_rate'].value_counts()

In [ ]:
off_test['discount_rate'].value_counts()

In [ ]:
off_train['distance'].value_counts()

In [ ]:
off_test['distance'].value_counts()

通过初步观察感觉训练集和测试集数据分布比较一致。

# 数据处理及分析

## 将特征数值化

In [109]:
#分隔符
separator=':'
#计算折扣率，将满减和折扣统一
#因为discount_rate为null的时候一般都是没有使用优惠券，这个时候折扣应该是1
def get_discount_rate(s):
    s = str(s)
    if s=='null':
        return -1
        #return 1
    s = s.split(separator)
    if len(s) == 1:
        return float(s[0])
    else:
        return 1.0-float(s[1])/float(s[0])

#获取是否满减（full reduction promotion）
def get_if_fd(s):
    s = str(s)
    s = s.split(separator)
    if len(s)==1:
        return 0
    else:
        return 1
        
#获取满减的条件
def get_full_value(s):
    s = str(s)
    s = s.split(separator)
    if len(s)==1:
        return -1
    else:
        return int(s[0])
        
#获取满减的优惠     
def get_reduction_value(s):
    s = str(s)
    s = s.split(separator)
    if len(s) == 1:
        return -1
    else:
        return int(s[1])


#获取月份
def get_month(s):
    if s[0]=='null':
        return -1
    else:    
        return int(s[4:6])

#获取日期
def get_day(s):
    if s[0]=='null':
        return -1
    else:    
        return int(s[6:8])
    
#获取日期间隔输入内容为Date:Date_received
def get_day_gap(s):
    s = s.split(separator)
    if s[0]=='null':
        return -1
    if s[1]=='null':
        return -1
    else:    
        return (date(int(s[0][0:4]),int(s[0][4:6]),int(s[0][6:8])) - date(int(s[1][0:4]),int(s[1][4:6]),int(s[1][6:8]))).days

#获取Label，输入内容为Date:Date_received
def get_label(s):
    s = s.split(separator)
    if s[0]=='null':
        return 0
    if s[1]=='null':
        return -1
    elif (date(int(s[0][0:4]),int(s[0][4:6]),int(s[0][6:8]))-date(int(s[1][0:4]),int(s[1][4:6]),int(s[1][6:8]))).days<=15:
        return 1
    else:
        return -1

In [110]:
def add_feature(df):
    df['if_fd']=df['discount_rate'].apply(get_if_fd)
    df['full_value']=df['discount_rate'].apply(get_full_value)
    df['reduction_value']=df['discount_rate'].apply(get_reduction_value)
    df['discount_rate']=df['discount_rate'].apply(get_discount_rate)
    df['distance']=df['distance'].replace('null',-1).astype(int)
    #df['month_received'] = df['date_received'].apply(get_month)
    #df['month'] = df['date'].apply(get_month)
    return df
    
def add_label(df):
    df['day_gap']=df['date'].astype('str') + ':' +  df['date_received'].astype('str')
    df['label']=df['day_gap'].apply(get_label)
    df['day_gap']=df['day_gap'].apply(get_day_gap)
    return df

In [111]:
#拷贝数据，免得调试的时候重读文件
dftrain = off_train.copy()
dftest = off_test.copy()

In [112]:
dftrain=add_feature(dftrain)
dftrain=add_label(dftrain)
dftest=add_feature(dftest)

## 数据分析

In [ ]:
dftrain.head()

In [ ]:
dftrain.describe()

In [ ]:
dftest.head()

In [ ]:
dftest.describe()

In [ ]:
dftrain[dftrain.distance>=0]['distance'].value_counts()/dftrain[dftrain.distance>=0]['distance'].count()

In [ ]:
dftest[dftest.distance>=0]['distance'].value_counts()/dftest[dftest.distance>=0]['distance'].count()

In [ ]:
dftrain[(dftrain.label>=0)&(dftrain.distance>=0)]['distance'].value_counts()/dftrain[(dftrain.label>=0)&(dftrain.distance>=0)]['distance'].count()

In [ ]:
dftest[(dftest.distance>=0)]['distance'].value_counts()/dftest[(dftest.distance>=0)]['distance'].count()

In [ ]:
print ('线下 训练集满减情况')
dftrain.if_fd.value_counts()/dftrain.if_fd.count()

In [ ]:
print ('测试集满减情况')
dftest.if_fd.value_counts()/dftest.if_fd.count()

## 箱线图查看分布

In [ ]:
fig = plt.figure(figsize=(4, 6))  # 指定绘图对象宽度和高度
sns.boxplot(dftrain[(dftrain.label>=0)&(dftrain.distance>=0)]['distance'],orient="v", width=0.5)

In [ ]:
fig = plt.figure(figsize=(4, 6))  # 指定绘图对象宽度和高度
sns.boxplot(dftrain[(dftrain.label>=0)&(dftrain.discount_rate>=0)]['discount_rate'],orient="v", width=0.5)

## 直方图和QQ图

In [ ]:
plt.figure(figsize=(10,5))

ax=plt.subplot(1,2,1)
sns.distplot(dftrain[(dftrain.label>=0)&(dftrain.distance>=0)]['distance'],fit=stats.norm)
ax=plt.subplot(1,2,2)
res = stats.probplot(dftrain[(dftrain.label>=0)&(dftrain.distance>=0)]['distance'], plot=plt)

In [ ]:
plt.figure(figsize=(10,5))

ax=plt.subplot(1,2,1)
sns.distplot(dftrain[(dftrain.label>=0)&(dftrain.discount_rate>=0)]['discount_rate'],fit=stats.norm)
ax=plt.subplot(1,2,2)
res = stats.probplot(dftrain[(dftrain.label>=0)&(dftrain.discount_rate>=0)]['discount_rate'], plot=plt)

## 对比分布

In [ ]:
ax = sns.kdeplot(dftrain[(dftrain.label>=0)&(dftrain.discount_rate>=0)]['discount_rate'], color="Red", shade=True)
ax = sns.kdeplot(dftest[(dftest.discount_rate>=0)]['discount_rate'], color="Blue", shade=True)
ax.set_xlabel('discount_rate')
ax.set_ylabel("Frequency")
ax = ax.legend(["train","test"])

In [ ]:
ax = sns.kdeplot(dftrain[(dftrain.label>=0)&(dftrain.distance>=0)]['distance'], color="Red", shade=True)
ax = sns.kdeplot(dftest[(dftest.distance>=0)]['distance'], color="Blue", shade=True)
ax.set_xlabel('distance')
ax.set_ylabel("Frequency")
ax = ax.legend(["train","test"])

In [ ]:
ax = sns.kdeplot(dftrain[(dftrain.label>=0)&(dftrain.full_value>=0)]['full_value'], color="Red", shade=True)
ax = sns.kdeplot(dftest[(dftest.full_value>=0)]['full_value'], color="Blue", shade=True)
ax.set_xlabel('full_value')
ax.set_ylabel("Frequency")
ax = ax.legend(["train","test"])

In [ ]:
ax = sns.kdeplot(dftrain[(dftrain.label>=0)&(dftrain.reduction_value>=0)]['reduction_value'], color="Red", shade=True)
ax = sns.kdeplot(dftest[(dftest.reduction_value>=0)]['reduction_value'], color="Blue", shade=True)
ax.set_xlabel('reduction_value')
ax.set_ylabel("Frequency")
ax = ax.legend(["train","test"])

## 可视化线性关系

In [ ]:
fcols = 2
frows = 1
plt.figure(figsize=(8,4))
ax=plt.subplot(1,2,1)
sns.regplot(x='distance', y='label', data=dftrain[(dftrain.label>=0)&(dftrain.distance>=0)][['distance','label']], ax=ax, 
            scatter_kws={'marker':'.','s':3,'alpha':0.3},
            line_kws={'color':'k'});
plt.xlabel('distance')
plt.ylabel('label')
ax=plt.subplot(1,2,2)
sns.distplot(dftrain[(dftrain.label>=0)&(dftrain.distance>=0)]['distance'].dropna())
plt.xlabel('distance')
plt.show()

In [ ]:
fcols = 2
frows = 1
plt.figure(figsize=(8,4))
ax=plt.subplot(1,2,1)
sns.regplot(x='discount_rate', y='label', data=dftrain[(dftrain.label>=0)&(dftrain.discount_rate>=0)][['discount_rate','label']], ax=ax, 
            scatter_kws={'marker':'.','s':3,'alpha':0.3},
            line_kws={'color':'k'});
plt.xlabel('discount_rate')
plt.ylabel('label')
ax=plt.subplot(1,2,2)
sns.distplot(dftrain[(dftrain.label>=0)&(dftrain.discount_rate>=0)]['discount_rate'].dropna())
plt.xlabel('discount_rate')
plt.show()